draft-ietf-dnsop-svcb-https.md

```diff
@@ -1082,18 +1082,31 @@ single-client granularity.
 
 ## Interaction with Alt-Svc
 
-Clients that do not implement support for Encrypted ClientHello MAY
-skip the HTTPS RR query
-if a usable Alt-Svc value is available in the local cache.
-If Alt-Svc connection fails, these clients SHOULD fall back to the HTTPS RR
-client connection procedure ({{client-behavior}}).
-
-Clients that implement support for ECH MUST perform the HTTPS RR query first,
-and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see
-{{ech-client-behavior}}).
-
-This specification does not alter the DNS records used when connecting
-to an Alt-Svc hostname (typically A and/or AAAA only).
+Clients that implement support for both Alt-Svc and HTTPS records SHOULD
+retrieve any HTTPS records for the Alt-Svc alt-authority, and ensure that
+their connection attempts are consistent with both the Alt-Svc parameters
+and any received HTTPS SvcParams.  If present, the HTTPS record's target
+and port override the alt-authority.  For example, suppose
+"https://example.com" sends an Alt-Svc field value of:
+
+    Alt-Svc: h3="alt.example.com:8443"; ...
+
+The client would retrieve the following HTTPS record:
+
+    _8443._https.alt.example.com. ... 1 alt2.example.com. (
+        port=9443 alpn=h2,h3 ech=... )
+
+The client could then attempt an HTTP/3 connection to `alt2.example.com:9443`
+with ECH, as this is consistent with both the Alt-Svc field value and the
+HTTPS record.  This specification does not alter or clarify the
+interpretation of Alt-Svc's `protocol-id` field.
+
+Origins that publish an "ech" SvcParam in their HTTPS record SHOULD
+also publish an "ech" SvcParam for any Alt-Svc hostnames.  Otherwise,
+clients may reveal the unencrypted ClientHello during Alt-Svc connections.
```

Comment:
Is it the case that the client will use the alternative service without ECH, or that it won't use the alternative service?

Or do we want to leave that unspecified?  I know that we allow some HTTPS/SVCB records to include ECH and some to omit it and that client policy governs which are chosen.  Is this just another case where client policy applies?

Personally, I would very much prefer saying that if you use ECH to connect, you MUST NOT use an alternative that doesn't have an ECH config.  "SHOULD NOT" would be acceptable on the grounds that we aren't dictating client policy.  But it's a pretty bad leak, and one that could be the result of a mistake in server configuration.

Reply:
> Or do we want to leave that unspecified?

FWIW, that's what I mean in this text (equivalent to normative MAY).

> Personally, I would very much prefer saying that if you use ECH to connect, you MUST NOT use an alternative that doesn't have an ECH config.

A client that wanted to implement that behavior would have to query _both_ the origin and the alt-authority for HTTPS records before it could proceed with an Alt-Svc connection.  If we want clients to take responsibility for ensuring ECH isn't lost, I think #327 makes more sense.

P.S. Typo: @carrickdb

Reply:
> > Personally, I would very much prefer saying that if you use ECH to connect, you MUST NOT use an alternative that doesn't have an ECH config.
> 
> A client that wanted to implement that behavior would have to query _both_ the origin and the alt-authority for HTTPS records before it could proceed with an Alt-Svc connection. If we want clients to take responsibility for ensuring ECH isn't lost, I think #327 makes more sense.

I think we're sufficiently covered due to the assumption (supported by a SHOULD here in the text) that if an origin publishes ECH for the origin endpoints, it also publishes ECH for the Alt-Svc endpoints.  Means that with a single DNS query, the client can find whether the server supports ECH unless the domains are stupidly configured (a similar level of stupidity as just forgetting to publish ECH keys on any other endpoint).

The current requirement is already that only if all endpoints found have ECH, the client MUST use the the ECH endpoints.  If a server is misconfigured to give non-ECH endpoints, I think it's best left to client choice how to prioritize ECH vs other factors like h3 (e.g. a non-privacy-sensitive client may choose to prioritize using an h3 endpoint if only h2 endpoints have ECH, or an extra-privacy-sensitive client may choose to not allow Alt-Svc at all unless ECH endpoints are found or may want to take the performance hit to query all Alt-Svc names and the origin to find any possible ECH endpoints before connecting to anything).

No matter what the client chooses as endpoint priorities, the proper configuration for an origin to force ECH is clear (and clearly recommended by section 9.2): Include ECH on all endpoints and section 9.1 requires ECH-capable clients to only connect via those endpoints.  This just now would apply to endpoints given on Alt-Svc names too.